apiserver/facades/client/action/operation.go

```diff
@@ -1,98 +1,279 @@
+	}
+
+	const leader = "/leader"
+	actionResults := make([]params.ActionResult, len(arg.Actions))
+	actionResultByUnitName := make(map[string]*params.ActionResult)
+	receivers := make([]operation.ActionReceiver, 0, len(arg.Actions))
+
+	taskParams := taskParamHolder{}
+	for i, action := range arg.Actions {
+		if err := taskParams.Set(action); err != nil {
+			return params.EnqueuedActions{},
+				errors.Errorf("all actions parameters from the same operation should be the same: %v", err)
+		}
+
+		if strings.HasSuffix(action.Receiver, leader) {
+			receiver := strings.TrimSuffix(action.Receiver, leader)
+			actionResultByUnitName[action.Receiver] = &actionResults[i]
+			receivers = append(receivers, operation.ActionReceiver{LeaderUnit: receiver})
+			continue
+		}
+
+		unitTag, err := names.ParseUnitTag(action.Receiver)
+		if err != nil {
+			actionResults[i].Error = apiservererrors.ServerError(err)
+			continue
+		}
+		actionResultByUnitName[unitTag.Id()] = &actionResults[i]
+		receivers = append(receivers, operation.ActionReceiver{Unit: unit.Name(unitTag.Id())})
+	}
+
+	// If no valid receivers (all are invalid), do not call service; return per-action errors.
+	if len(receivers) == 0 {
+		return params.EnqueuedActions{Actions: actionResults}, nil
+	}
+
+	result, err := a.operationService.StartActionOperation(ctx, receivers, taskParams.Get())
 	if err != nil {
-		return params.EnqueuedActions{}, err
+		return params.EnqueuedActions{}, errors.Capture(err)
+	}
+
+	for _, unitResult := range result.Units {
+		targetName := unitResult.ReceiverName.String()
+		if unitResult.IsLeader {
+			targetName = unitResult.ReceiverName.Application() + leader
+		}
+		ar := actionResultByUnitName[targetName]
+		if ar == nil {
+			return params.EnqueuedActions{}, errors.Errorf("unexpected result for %q", targetName)
+		}
+		*ar = toActionResult(names.NewUnitTag(unitResult.ReceiverName.String()), unitResult.TaskInfo)
 	}
-	results := params.EnqueuedActions{
-		OperationTag: names.NewOperationTag(operationId).String(),
-		Actions:      actionResults.Results,
+	// Mark missing results
+	for key, ar := range actionResultByUnitName {
+		if ar != nil && ar.Action == nil && ar.Error == nil {
+			ar.Error = apiservererrors.ServerError(errors.Errorf("missing result for %q", key).Add(coreerrors.NotFound))
+		}
 	}
-	return results, nil
+	return params.EnqueuedActions{
+		OperationTag: names.NewOperationTag(result.OperationID).String(),
+		Actions:      actionResults,
+	}, nil
 }
 
-func (a *ActionAPI) enqueue(ctx context.Context, arg params.Actions) (string, params.ActionResults, error) {
-	if err := a.checkCanWrite(ctx); err != nil {
-		return "", params.ActionResults{}, errors.Trace(err)
+// taskParamHolder is a parameter holder for the task parameters.
+// It helps to validate that all parameters for bulk actions from the facade
+// are the same.
+type taskParamHolder struct {
```

Comment:
This struct's name and methods don't represent correctly what it does IMHO. Please correct me if I'm wrong, this is only used to validate the params, making sure that all the common parameters of the bulk inputs are exactly the same.
I suggest having either a `Validate(actions params.Actions)`, in which you simply iterate thorough all the bulk inputs and make sure the params are the same, like you do in the `Set()` method.
Either having some databag struct, but instead of setter/getter just have a `Validate()` method, like we do with the manifold configs for example.